File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;

/*알고리즘
- DFS로 단어 생성
    - 빈 문자열 -> "A", "E", "I", "O", "U"를 하나씩 추가해 단어 만들기
    - 최대 5글자(count 증가)
- 목표 단어(word)를 찾으면 -> 종료
    - 생성한 단어가 word와 같다면 -> answer를 count 값으로
- word를 찾으면 바로 return
*/

class Solution {
    static String[] words = {"A", "E", "I", "O", "U"};
    static int count, answer;
    
    public int solution(String word) {
        count = 0;
        answer = 0;
        dfs("", word, 0);
        return answer;
    }

    static void dfs(String str, String target, int len) {
        if (str.equals(target)) {
            answer = count;
            return;
        }
        if (len == 5) return;
        
        for (int i = 0; i < 5; i++) {
            count++;
            dfs(str + words[i], target, len + 1);
            if (answer > 0) return;
        }
    }
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
# 알고리즘
# 0. 모음 사전 형태 이해
# A
# AA
# AAA
# AAAA
# AAAAA
# AAAAE
#...
# UUUUU
# 1. 알파벳 모음 사전 배열 선언, 단어 모음 words 배열 선언
# 2. DFS 함수
# 2-1. 길이가 5 초과면 탐색 중단 (문제 조건)
# 2-2. 빈 문자열은 제외하고 저장
# 2-3. # 각 모음을 하나씩 붙여서 재귀 호출
# 3. DFS 실행
# 3-1. 빈 문자열부터 시작 (모든 조합을 탐색)
# 4. 사전순 정렬(words)
# 5. 위치는 0부터 시작하므로 +1

def solution(word):
    alpha = ['A', 'E', 'I', 'O', 'U']
    words = []
    
    def dfs(curr):
        if len(curr) > 5:
            return
        if curr:
            words.append(curr)
        for ch in alpha:
            dfs(curr + ch)
    
    dfs("")
    words.sort()
    
    answer = words.index(word) + 1
    return answer